drivers: display: reimplementation of micro:bit display

peterniebert opened the pull request:
Reimplementation of LED matrix driver for micro:bit. Here is why:

- It is compatible both with micro:bit V1 and V2
- It improves the rendering by rendering pixels rather than lines (the latter causes uneven luminosity), while maintaining a low system load.

This is achieved by using a hardware timer of the SOC for pixel refresh interrupts and by an optimized pixel refresh ISR. The estimated maximum load while the screen is active is less than 1%, maybe 0.5%. The timing of pixel refreshing and animation is separated. We use delayable kwork for the animation.

The driver is rewritten from sketch. I have tested it with both micro:bit V1 and V2. It's behaviour should be very similar to the original driver on V1.

So far, the only sample I have adapted is samples/boards/bbc_microbit/display. The only change needed was to require an NRFX_TIMER to be enabled. I will go through the other demos on V1, the trouble is, some of them still won't work on V2 for other reasons.

Happy to receive your comments.

aaillet commented:
Hi, 
I think that you made a mistake while rebasing your work.
You squashed your work in a commit that has just been merged in main instead of rebasing on the top of the main branch !
You should maybe rework your commit tree !

peterniebert commented:
> Hi,
> I think that you made a mistake while rebasing your work.
> You squashed your work in a commit that has just been merged in main instead of rebasing on the top of the main branch (your commit is then also containing the work from this commit )!
> You should maybe rework your commit tree !

Totally messed it up! I will work on it. It is going to be fun.

peterniebert committed:
drivers: display: reimplementation of micro:bit display

Reimplementation of LED matrix driver for micro:bit, formally
drivers/display/mb_display.c

The new driver adds compatibility with micro:bit v2. Also,
it improves the rendering by rendering pixels rather than
lines (the latter caused uneven luminosity), while maintaining
a low system load.

Signed-off-by: Peter Niebert <peter.niebert@univ-amu.fr>

peterniebert commented:
> > Hi,
> > I think that you made a mistake while rebasing your work.
> > You squashed your work in a commit that has just been merged in main instead of rebasing on the top of the main branch (your commit is then also containing the work from this commit )!
> > You should maybe rework your commit tree !
> 
> Totally messed it up! I will work on it. It is going to be fun.
It is OK now, looking

> > Hi,
> > I think that you made a mistake while rebasing your work.
> > You squashed your work in a commit that has just been merged in main instead of rebasing on the top of the main branch (your commit is then also containing the work from this commit )!
> > You should maybe rework your commit tree !
> 
> Totally messed it up! I will work on it. It is going to be fun.

It seems OK now, looking forward to your comments. My guess is that there will be some discussion, so a few words to explain the state of the suggested commit.

- it preserves backward compatibility with the original API, the sample/boards/bbc_microbit/display works - almost - unchanged.
- what is changed is that the new driver uses a NRF hardware timer. It can be chosen in order to resolve conflicts, but so the project file must activate the chosen timer. I don't understand Kconfig enough to make this more automatic, maybe somebody has a suggestion.

- Before opting for pixel based refresh (which means 2500 interrupts per second), I tried line based multiplexing. As expected, the line based multiplexing has a problem with the parallel current in case of several LEDs in the same line active. To make it short, as the current drain rises there is a voltage drop and this results in lower luminosity.  It turns out that with a very small ISR, the system load resulting from the refresh is something like 0.4%. It also turns out that it is much easier to write modular code that works for both matrices (9x3 for V1 and 5x5 for V2). A single table encoding the two pins connecting each LED and representing the topology is sufficient.

There are a few improvements possible, at least for micro:bit v2: For instance, either cathode or anode could be substituted by a PWM output. As a result, one might without additional cost change the brightness. Also, if someone should find the brightness too low, it is possible to change from standard drive to high drive. Inspite of pixel oriented refresh, with high drive the display is blindingly bright.

- The final topic concerns the API. The original API has a completely useless argument to all functions, the driver object. It is clearly preferable to have it statically allocated, but then why pass it around. So right now, I put two header files, one for compatibility, the other without these arguments. Otherwise, I think that the API does what is expected of it.

Thank you for your opinion.

-- Peter

peterniebert commented:
It turned out that some samples did not compile, so I changed the configuration process to automate the NRFX_TIMERn activation.

peterniebert commented:
latex failing is not my fault, I guess?

peterniebert commented:
Sorry for all the force-pushes, I had to get rid of the failing tests, finally it passes :-)

BTW, in a different pull request it has been suggested to move this driver out of display to misc. I can do that if you wish.

The essential thing is to rapidly get a driver for the micro:bit v2 into zephyr as it seems to be an important "development card" and really an ideal object for teaching zephyr (this is my motivation for putting energy into this).

peterniebert committed:
drivers: display: move microbit led driver to drivers: misc

The mb_display.h interface does not implement any generic Zephyr
API. This patch moves and renames all related files to drivers/misc.
For compatibility with legacy code, the folder
drivers/misc/microbit_display is added a file display/mb_display.h
This respects the #include "display/mb_display.h" clauses found
in some examples. Once the examples are patched, the file can
receive a depreciation warning.

Signed-off-by: Peter Niebert <peter.niebert@univ-amu.fr>

peterniebert commented:
> Sorry for all the force-pushes, I had to get rid of the failing tests, finally it passes :-)
> 
> BTW, in a different pull request it has been suggested to move this driver out of display to misc. I can do that if you wish.

To see how this could look, I have added a second commit, which moves everything to drivers/misc.

peterniebert commented:
I'm lost on the bug in the documentation, is there any good way of debugging it?

tejlmand reviewed (changes requested):
Some change requests regardning Kconfig.

Also are there any samples or tests using this display ?

In order to verify that things are working.

hubertmis commented on drivers/misc/microbit_display/microbit_display.h:
```diff
@@ -1,20 +1,16 @@
 
-#ifndef ZEPHYR_INCLUDE_DISPLAY_MB_DISPLAY_H_
-#define ZEPHYR_INCLUDE_DISPLAY_MB_DISPLAY_H_
+#ifndef ZEPHYR_INCLUDE_DISPLAY_MB_DISPLAY_V2_H_
```
```suggestion
#ifndef ZEPHYR_INCLUDE_MISC_MICROBIT_DISPLAY_MICROBIT_DISPLAY_H_
```

carlescufi commented:
@peterniebert thank you very much for the PR! I have added some reviewers, would very much like to see this merged soon.

hubertmis commented on drivers/misc/microbit_display/Kconfig:
```diff
@@ -0,0 +1,93 @@
+
+config MICROBIT_DISPLAY_TIMER0
+		bool "TIMER0"
+		depends on HAS_HW_NRF_TIMER0
```
I think it would be easier to avoid conflicts if you used `TIMERX_RESERVED` config entries.
```suggestion
		depends on HAS_HW_NRF_TIMER0
		depends on !NRF_HW_TIMER0_RESERVED
```

gmarull commented on drivers/display/Kconfig.microbit:
```diff
@@ -1,23 +1,93 @@
+		default MICROBIT_DISPLAY_TIMER3
+		help
+			Select the hardware timer instance used by this driver.
+			Beware of conflicts with other drivers, e.g. for bluetooth.
```
can't this be selected somehow in DT?

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+	15 << 8 | 5 };
+
+static const uint32_t row_mask0 =
+	1 << 4 | 1 << 5 | 1 << 6 | 1 << 7 | 1 << 8 | 1 << 9 | 1 << 10 | 1 << 11 | 1 << 12;
```
this needs some better formatting

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+
+};
+
+/* TODO check if C compiler handles this efficiently */
```
?

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+#else
+#error "microbit led matrix only works on boards bbc_microbit and bbc_microbit_v2."
+#endif
+#endif
```
is it strictly necessary to use nrf API instead of the generic GPIO API?

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+};
+
+/* unique instance of this struct */
+static struct myscreen_struct myscreen;
```
why can't this API implement a regular device (with data, config, a init function, etc.)?

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+	const struct mb_image *img1;
+	const struct mb_image *img2 = NULL;
+	int scroll_offset;
+	int scrollsteps = 5;
```
constants need to be defined

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+		if (myscreen.scrolling) {
+			scrollsteps = 5;
+			index = myscreen.current_image / 5;
+			scroll_offset = myscreen.current_image - 5 * index;
```
ditto

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+
+		mypinclear(row);
+		nrf_gpio_port_out_clear(NRF_P0, col_mask);
+		int column = col_row >> 8;
```
again, undefined constants

gmarull commented on drivers/display/mb_display_v2.c:
```diff
@@ -0,0 +1,514 @@
+ */
+
+static int init_driver(const struct device *dev);
+SYS_INIT(init_driver, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
```
ditto: use a regular device

gmarull commented on include/display/mb_display.h:
```diff
@@ -11,161 +16,44 @@
+inline struct mb_display *mb_display_get(void)
+{
+	return NULL;
+}
```
what is the point of this function?

gmarull commented on include/display/mb_display.h:
```diff
@@ -11,161 +16,44 @@
+	{ \
+		(void) disp; \
+		mb_display_image_v2(mode, duration, img, img_count); \
+	}
```
why not `static inline` shim?

gmarull commented on include/display/mb_display.h:
```diff
@@ -11,161 +16,44 @@
+	{ \
+		(void) disp; \
+		mb_display_stop_v2(); \
+	}
```
ditto

gmarull reviewed (changes requested):
Done initial review. I think the code needs a bit of refactoring to align with Zephyr coding standards and device drivers structure.

hubertmis commented on drivers/misc/microbit_display/microbit_display.c:
```diff
@@ -0,0 +1,514 @@
+/* columns/anodes followed by rows/cathodes */
+
+static const uint16_t led_column_row[25] = {
+	21 << 8 | 28, 21 << 8 | 11, 21 << 8 | 31, 21 << 8 | 37, 21 << 8 | 30,
+	22 << 8 | 28, 22 << 8 | 11, 22 << 8 | 31, 22 << 8 | 37, 22 << 8 | 30,
+	15 << 8 | 28, 15 << 8 | 11, 15 << 8 | 31, 15 << 8 | 37, 15 << 8 | 30,
+	24 << 8 | 28, 24 << 8 | 11, 24 << 8 | 31, 24 << 8 | 37, 24 << 8 | 30,
+	19 << 8 | 28, 19 << 8 | 11, 19 << 8 | 31, 19 << 8 | 37, 19 << 8 | 30
```
Are these pin numbers? If so, would it be possible to define them in device tree, instead of source files?
